cabal-install/Distribution/Client/CmdRun.hs

```diff
@@ -206,23 +206,53 @@ runAction (configFlags, configExFlags, installFlags, haddockFlags)
                                   pkg
                                   exeName
                </> exeName
-    curDir <- getCurrentDirectory
-    let dataDirEnvVar = (pkgPathEnvVar (elabPkgDescription pkg) "datadir",
-                         Just $ curDir </> dataDir (elabPkgDescription pkg))
-        args = drop 1 targetStrings
+    let args = drop 1 targetStrings
     runProgramInvocation
       verbosity
       emptyProgramInvocation {
         progInvokePath  = exePath,
         progInvokeArgs  = args,
-        progInvokeEnv   = [dataDirEnvVar]
+        progInvokeEnv   = dataDirsEnvironmentForPlan elaboratedPlan
       }
   where
     verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
     cliConfig = commandLineFlagsToProjectConfig
                   globalFlags configFlags configExFlags
                   installFlags haddockFlags
 
+
+-- | Construct the environment needed for the data files to work.
+-- This consists of a separate @*_datadir@ variable for each
+-- inplace package in the plan.
+dataDirsEnvironmentForPlan :: ElaboratedInstallPlan
+                           -> [(String, Maybe FilePath)]
+dataDirsEnvironmentForPlan = catMaybes
+                           . fmap (foldPlanPackage
+                               (const Nothing)
+                               dataDirEnvVarForPackage)
+                           . toList
+
+-- | Construct an environment variable that points
+-- the package's datadir to its correct location.
+-- This might be:
+-- * 'Just' the package's source directory plus the data subdirectory
+--   for inplace packages.
+-- * 'Nothing' for packages installed in the store (the path was
+--   already included in the package at install/build time).
+-- * The other cases are not handled yet. See below.
+dataDirEnvVarForPackage :: ElaboratedConfiguredPackage
+                        -> Maybe (String, Maybe FilePath)
```

Comment:
That's the format that runProgramInvocation accepts. See `progInvokeEnv` in `Cabal/Distribution/Simple/Program/Run.hs`.

The inner Maybe is for valueless vars I think.

Oddly, we don't have an alias for it.